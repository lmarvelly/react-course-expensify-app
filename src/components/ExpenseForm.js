import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component
{	
	constructor(props)
	{
		super(props);

		// Sets up the default state if it's a new expense or sets 
		// up current values for editing expense
		this.state = 
		{
			// if an expense exists
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			note: props.expense ? props.expense.note : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calenderFocused: false,
			error: ''
		}
	}
	
	onDescriptionChange = (e) =>
	{
		const description = e.target.value;
		this.setState(() => ({ description }));
	}
	onAmountChange = (e) =>
	{
		const amount = e.target.value;

		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
		{
			this.setState(() => ({ amount }));
		}
	}
	onDateChange = (createdAt) =>
	{
		if (createdAt) // If statement is needed to prevent user from deleting date
		{
			this.setState(() => ({ createdAt }));
		}
	};
	onFocusChange = ({ focused }) =>
	{
		this.setState(() => ({ calenderFocused: focused }));
	};
	onNoteChange = (e) =>
	{
		const note = e.target.value;
		this.setState(() => ({ note }));
	}
	onSubmit = (e) =>
	{
		e.preventDefault(); // prevent full page reload

		if (!this.state.description || !this.state.amount)
		{
			this.setState(() => ({ error: 'Please provide description and amount.' }));
		}
		else
		{
			this.setState(() => ({ error: ''}));
			// onSubmit gets passed down by the AppExpensePage
			this.props.onSubmit(
			{
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
	}
	render()
	{
		return (
			<div>
				{ this.state.error && <p>{ this.state.error }</p>}
				<form onSubmit={ this.onSubmit }>
					<input
						type='text'
						placeholder='Description'
						autoFocus
						value={ this.state.description }
						onChange={ this.onDescriptionChange }
					/>
					<input
						type='text'
						placeholder='Amount'
						value={ this.state.amount }
						onChange={ this.onAmountChange }
					/>
					<SingleDatePicker 
						date={ this.state.createdAt }
						onDateChange={ this.onDateChange }
						focused= { this.state.calenderFocused }
						onFocusChange={ this.onFocusChange }
						numberOfMonths={ 1 }
						isOutsideRange={ () => false } // makes everysingle day in future and past available
						displayFormat="DD/MM/YYYY" // UK date format
					/>
					<textarea
						placeholder='Add a note for your expense (optional)'
						value={ this.state.note }
						onChange={ this.onNoteChange }
					></textarea>
					<button>Add Expense</button>
				</form>
			</div>
		)
	}
}