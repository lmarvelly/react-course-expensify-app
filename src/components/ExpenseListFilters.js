import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';

/**
 * @function ExpenseListFilters
 * 
 * @param {*} props 
 * 
 * @e This is the event target. 
 */
const ExpenseListFilters = (props) =>
(
	<div>
		<input 
			type='text' 
			value={ props.filters.text }
			onChange={(e) =>
			{
				props.dispatch(setTextFilter(e.target.value));
			}}
		/>
	</div>
);

const mapStateToProps = (state) =>
{
	return {
		filters: state.filters
	}
};

export default connect(mapStateToProps)(ExpenseListFilters);