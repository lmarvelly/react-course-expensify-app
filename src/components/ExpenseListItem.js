import React from 'react';
import { Link } from 'react-router-dom'

// If we want to see note then we add it as an argument and then you can add it to the paragraph: { note }
export const ExpenseListItem = ({ id, description, amount, createdAt }) => 
(
	<div>
		<Link to={ `/edit/${id}` }><h3>{ description } </h3></Link>

		<p>{ amount } - { createdAt }</p>
	</div>
);	

export default ExpenseListItem;