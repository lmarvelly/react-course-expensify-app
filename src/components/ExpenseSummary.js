import React from 'react';
import selectExpensesTotal from '../selectors/expenses-total'; 
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpenseSummary = (props) =>
(
	<div>
		<p>Total expenses: { props.expenseTotal }</p>
		<p>Total spent: { numeral( props.expenses / 100 ).format('$0,0.00') }</p>
	</div>
);

const mapStateToProps = (state) =>
{
	return { 
		expenses: selectExpensesTotal( state.expenses ),
		expenseTotal: state.expenses.length
	}
}

export default connect(mapStateToProps)(ExpenseSummary);