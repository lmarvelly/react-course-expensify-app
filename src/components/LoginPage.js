import React from 'react';

const LoginPage = () =>
(
	<div>
		<form>
			<input type='text' placeholder='Username' />
			<input type='password' placeholder='Password' />
		</form>
	</div>
);

export default LoginPage;