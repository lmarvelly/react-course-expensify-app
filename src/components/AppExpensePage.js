import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expensers'

const AddExpensePage = (props) =>
(
	<div>
		<h1>Add Expense</h1>
		<ExpenseForm
			onSubmit={(expense) => 
			{
				props.dispatch(addExpense(expense));
				props.history.push('/'); // redirect to dashboard
			}}
		/>
	</div>
);

export default connect()(AddExpensePage); // 