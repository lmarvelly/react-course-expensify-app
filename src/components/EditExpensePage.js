import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


export class EditExpensePage extends React.Component
{
	onSubmit = (expense) => 
	{
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};
	onRemove = () =>
	{
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	}
	render()
	{
		return(
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm
						expense={ this.props.expense }
						onSubmit={ this.onSubmit }
					/>
					<button 
						className="button button--secondary" 
						onClick={ this.onRemove }
					>
						Remove Expense
					</button>
				</div>
			</div>
		);
	}
}

// maps the Store state to the component props when passed into connect() 
// We need this to access the expense we want to edit. It returns the expense we need
const mapStateToProps = (state, props) =>
({
	/**
	 * props.match.params.id: the id that was passed in when 
	 * linking this page from the dashboard.
	 * When the above id matches an expense id, that expense will 
	 * be returned
	 */
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) =>
({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);