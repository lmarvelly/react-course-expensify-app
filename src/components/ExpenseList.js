import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// ExpenseList didn't need anything passed down because we imported
// and used connect()
const ExpenseList = (props) => (
	<div>
		<h1>Expense List</h1>
		{
			props.expenses.map((expense) => 
			{
				return <ExpenseListItem key={ expense.id } { ...expense } />
			})
		}
	</div>
);

// Function that maps the Store state to the component props
const mapStateToProps = (state) =>
{
	return {
		expenses: selectExpenses( state.expenses, state.filters )
	};
};

// 
export default connect(mapStateToProps)(ExpenseList);