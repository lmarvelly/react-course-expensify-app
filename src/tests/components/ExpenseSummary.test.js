import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';

// test('should generate an expense totaling Â£46.95', () =>
// {
// 	const wrapper = shallow(<ExpenseSummary expenses={ [] } />);
// 	expect(wrapper).toMatchSnapshot();
// });

test('should generate an expense totaling $2.35', () =>
{
	// Shallow render JSX
	const wrapper = shallow(<ExpensesSummary expenseCount={ 1 } expensesTotal={ 235 } />);
	expect(wrapper).toMatchSnapshot();
});

test('should generate an expense totaling $25.23', () =>
{
	const wrapper = shallow(<ExpensesSummary expenseCount={ 5 } expensesTotal={ 2523 } />);
	expect(wrapper).toMatchSnapshot();
});