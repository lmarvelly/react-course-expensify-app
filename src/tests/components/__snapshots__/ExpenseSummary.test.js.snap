// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate an expense totaling $2.35 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;

exports[`should generate an expense totaling $25.23 1`] = `
<div>
  <h1>
    Viewing 
    5
     
    expenses
     totalling 
    $25.23
  </h1>
</div>
`;
